addons/l10n_sa_edi/models/account_move.py

```diff
@@ -0,0 +1,673 @@
+        """
+            Force recalculate unsigned data (UUID, XML content, signature) for the invoice in case it is needed.
+        """
+        return self._l10n_sa_generate_unsigned_data()
+
+    def _l10n_sa_generate_unsigned_data(self):
+        """
+            Generate Unsigned XML content & digital signature to be used during both Signing and QR code generation.
+            It is necessary to save the signature as it changes everytime it is generated and both the signing and the
+            QR code expect to have the same, identical signature.
+        """
+        self.ensure_one()
+        edi_format = self.env.ref('l10n_sa_edi.edi_sa_zatca')
+        # Build the dict of values to be used for generating the Invoice XML content
+        # Set Invoice field values required for generating the XML content, hash and signature
+        self.l10n_sa_uuid = uuid.uuid4()
+        self.l10n_sa_unsigned_xml = edi_format._l10n_sa_generate_zatca_template(self)
+        # Once the required values are generated, we hash the invoice, then use it to generate a Signature
+        invoice_hash_hex = self.env['account.edi.xml.ubl_21.zatca']._l10n_sa_generate_invoice_xml_hash(
+            self.l10n_sa_unsigned_xml).decode()
+        self.l10n_sa_invoice_signature = edi_format._l10n_sa_get_digital_signature(self.journal_id.company_id,
+                                                                                   invoice_hash_hex).decode()
+
+    def _l10n_sa_save_submission(self, xml_content, invoice_name, response_data=None):
+        """
+            Save submitted invoice XML content in case of either Rejection or Acceptance.
+            This is useful to figure out why an invoice was rejected since we also save the errors along with the state.
+        """
+        self.ensure_one()
+        is_error = response_data and response_data.get('error')
+        attachment_id = self.env['ir.attachment'].create({
+            'name': invoice_name,
+            'raw': xml_content,
+            'res_model': 'account.move',
+            'res_id': self.id,
+            'mimetype': 'application/xml'
+        })
+        self.l10n_sa_submission_ids = [(0, 0, {
+            "l10n_sa_move_id": self.id,
+            "l10n_sa_chain_index": self.l10n_sa_chain_index,
+            "l10n_sa_attachment_id": attachment_id.id,
+            "l10n_sa_status": "rejected" if is_error else "accepted",
+            "l10n_sa_errors": is_error,
+        })]
+        return attachment_id
+
+    def _l10n_sa_ensure_line_taxes(self):
+        """
+            All invoice lines submitted to ZATCA need to have at least one Tax applied
+        """
+        self.ensure_one()
+        return all(line.tax_ids for line in self.invoice_line_ids.filtered(lambda line: not line.display_type))
+
+
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    @api.model
+    def _get_query_tax_details(self, tables, where_clause, where_params, fallback=True):
```

Comment:
Why does not a simple filtering work here?  (there is a param for it)

Reply:
Done